lib/models/event.dart

```diff
@@ -55,160 +47,76 @@ class Event {
-    _recurrenceRules = newRecurrence;
-  }
-
-  Recurrence get recurrenceRules => _recurrenceRules;
-
-  set timeStart(DateTime newTimeStart) {
-    _timeModified = DateTime.now();
-    _timeStart = newTimeStart;
+  Event.clone(Event e)
+      : this(
+            name: e.name,
+            id: e.id,
+            description: e.description,
+            color: e.color,
+            location: e.location,
+            tags: e.tags,
+            recurrenceRules: e.recurrenceRules,
+            timeStart: e.timeStart,
+            timeEnd: e.timeEnd,
+            timeCreated: e.timeCreated,
+            timeModified: e.timeModified);
+
+  /// returns a mapping with kv pairs corresponding to Firebase's
+  /// possibly a better getter
+  @override
+  Map<String, dynamic> toMap({keepClasses = false, includeID = false}) {
+    Map<String, dynamic> map = super.toMap();
+    map['time end'] = timeEnd;
+    return map;
   }
 
-  DateTime get timeStart => _timeStart;
-
   set timeEnd(DateTime newTimeEnd) {
-    _timeModified = DateTime.now();
-    _timeEnd = newTimeEnd;
+    timeModified = DateTime.now();
+    timeEnd = newTimeEnd;
   }
 
   DateTime get timeEnd => _timeEnd;
 
-  // Do not want to timeCreated this after the constructor
-  get timeCreated => _timeCreated; 
-
-  // Do not want to change timeModified unless modifying a field
-  get timeModified => _timeModified; 
-
-  int _daysBetween(DateTime from, DateTime to) {
-    from = DateTime(from.year, from.month, from.day);
-    to = DateTime(to.year, to.month, to.day);
-    return (to.difference(from).inHours / 24).round();
-  }
-
   /// Some checks to make sure the event object is valid with recurrence and crash with a more useful error message if caught
   /// used internally to run some basic checks before I assume things aren't null/are valid classes
   bool _validEventWithRecurrence() {
-
     // event fields must be valid, crash since something has probably gone very wrong
     if (timeStart == null || timeEnd == null) {
       throw Exception("Event is malformed? No timeStart/timeEnd value is set!");
     }
 
     // recurrence must have its fields be populated
-    if (recurrenceRules!.timeStart == null || recurrenceRules!.timeEnd == null || recurrenceRules!.dates == null) {
+    if (recurrenceRules!.timeStart == null ||
+        recurrenceRules!.timeEnd == null ||
+        recurrenceRules!.dates == null) {
       return false;
     }
-    if (recurrenceRules?.timeStart == null || recurrenceRules?.timeEnd == null) {
-      throw Exception("Event recurrence rules are enabled and not null, but the rest of the recurrence rules fields are unset!"); // I think this should be enforced
+    if (recurrenceRules?.timeStart == null ||
+        recurrenceRules?.timeEnd == null) {
+      throw Exception(
+          "Event recurrence rules are enabled and not null, but the rest of the recurrence rules fields are unset!"); // I think this should be enforced
     }
     return true;
   }
```

Comment:
Not sure if this is still needed. All the instance variables of recurrence are not nullable anymore, but I think that may need to change soon. 

Not sure how recurrence currently works, is it something like:
* if recurrence.enabled is true, then we are guaranteed correct recurrence start and end times.
* if recurrence.enabled is false, then we don't look at anything else
?

@stxue1

Reply:
my IDE ran into the same thing too. In the local branch I'm working on I just removed all instances so the IDE would stop complaining.

Reply:
I think we should always ensure correct recurrence start and end times since we're going towards all datetime variables being not nullable.